rotator/rotator.go

```diff
@@ -47,13 +47,15 @@ var nl = []byte{'\n'}
 // A Rotator writes input to a file, splitting it up into gzipped chunks once
 // the filesize reaches a certain threshold.
 type Rotator struct {
-	size      int64
-	threshold int64
-	maxRolls  int
-	filename  string
-	out       *os.File
-	tee       bool
-	wg        sync.WaitGroup
+	size           int64
+	threshold      int64
+	maxRolls       int
+	filename       string
+	out            *os.File
+	tee            bool
+	newCompressor  func(io.Writer) io.WriteCloser
```

Comment:
I think this should support a fallible compressor constructor / also return an error. E.x. if a user wants to still use gzip, but with a different compression level:

https://pkg.go.dev/compress/gzip@go1.23.0#NewWriterLevel

That could return an error if the compression level is invalid. Same for zstd:

https://pkg.go.dev/github.com/klauspost/compress/zstd#NewWriter

Suggested signature:

`func(io.Writer) (io.WriteCloser, error)`

Reply:
I saw that, but the proposed API for the stdlib doesn't return an error.

https://github.com/klauspost/compress/blob/v1.17.9/zstd/encoder.go#L66

It appears to only error for invalid options, which is probably something that could reasonably be panic'd on instead.

Reply:
I suppose I'll change it to pass the error through, since it doesn't make the default use case any more complex.

Reply:
Hmm true, but the existing stdlib compressors (like gzip) do have the `(io.WriteCloser, error)` return whenever using custom options.

I think for zstd they are suggesting to have a separate `SetLevel(int)` method instead of a fallible `NewWriter()` method.

https://github.com/golang/go/issues/62513#issuecomment-2289360138

`flate` from stdlib also uses this pattern:

https://pkg.go.dev/compress/flate@go1.23.0#NewWriterDict

I'd rather avoid panics on a bad compression level option, and if you were using an infallible constructor like `gzip.NewWriter()` you would just need a `, nil` at the end to use my suggested pattern.